go/kbfs/kbfsedits/tlf_history_test.go

```diff
@@ -427,9 +429,131 @@ func TestTlfHistoryRenameParentSimple(t *testing.T) {
+	require.NoError(t, err)
+
+	var aliceMessages, bobMessages []string
+	nn := nextNotification{1, 0, tlfID, nil}
+
+	// Alice and bob each delete one file, then create different files.
+	aliceDeleteA := nn.make("a", NotificationDelete, aliceUID, nil, time.Time{})
+	aliceMessages = append(aliceMessages, nn.encode(t))
+	bobDeleteB := nn.make("b", NotificationDelete, bobUID, nil, time.Time{})
+	bobMessages = append(bobMessages, nn.encode(t))
+
+	aliceWrite := nn.make("c", NotificationCreate, aliceUID, nil, time.Time{})
+	aliceMessages = append(aliceMessages, nn.encode(t))
+	bobWrite := nn.make("d", NotificationCreate, bobUID, nil, time.Time{})
+	bobMessages = append(bobMessages, nn.encode(t))
+
+	expected := writersByRevision{
+		{bobName,
+			[]NotificationMessage{bobWrite},
+			[]NotificationMessage{bobDeleteB},
+		},
+		{aliceName,
+			[]NotificationMessage{aliceWrite},
+			[]NotificationMessage{aliceDeleteA},
+		},
+	}
+
+	// Alice, then Bob.
+	th := NewTlfHistory()
+	rev, err := th.AddNotifications(aliceName, aliceMessages)
+	require.NoError(t, err)
+	require.Equal(t, aliceWrite.Revision, rev)
+	rev, err = th.AddNotifications(bobName, bobMessages)
+	require.NoError(t, err)
+	require.Equal(t, bobWrite.Revision, rev)
+	checkTlfHistory(t, th, expected, aliceName)
+
+	// Another delete from alice, which should change the order of the
+	// expected history (bob should still come first).
+	aliceDeleteE := nn.make("e", NotificationDelete, aliceUID, nil, time.Time{})
+	aliceMessages = append(aliceMessages, nn.encode(t))
+	expected[1].deletes = []NotificationMessage{aliceDeleteE, aliceDeleteA}
+	rev, err = th.AddNotifications(aliceName, aliceMessages)
+	require.NoError(t, err)
+	require.Equal(t, aliceDeleteE.Revision, rev)
+	checkTlfHistory(t, th, expected, aliceName)
+
+	// Now add > 10 writes each, to make sure the deletes remain in
+	// the history.
+	var allAliceExpected, allBobExpected notificationsByRevision
+	for i := 0; i < 2*(maxEditsPerWriter+1); i += 2 {
+		event := nn.make(
+			strconv.Itoa(i), NotificationCreate, aliceUID, nil, time.Time{})
+		allAliceExpected = append(allAliceExpected, event)
+		aliceMessages = append(aliceMessages, nn.encode(t))
+		event = nn.make(
+			strconv.Itoa(i+1), NotificationCreate, bobUID, nil, time.Time{})
+		allBobExpected = append(allBobExpected, event)
+		bobMessages = append(bobMessages, nn.encode(t))
+	}
+	sort.Sort(allAliceExpected)
+	sort.Sort(allBobExpected)
+
+	expected[0].notifications = allBobExpected[:maxEditsPerWriter]
+	expected[1].notifications = allAliceExpected[:maxEditsPerWriter]
+	rev, err = th.AddNotifications(aliceName, aliceMessages)
+	require.NoError(t, err)
+	require.Equal(t, allAliceExpected[0].Revision, rev)
+	rev, err = th.AddNotifications(bobName, bobMessages)
+	require.NoError(t, err)
+	require.Equal(t, allBobExpected[0].Revision, rev)
+	checkTlfHistory(t, th, expected, aliceName)
+
+	// Re-creating a deleted file should remove the delete.
```

Comment:
If I were looking in my edit history and I had changed something, reverted it, and then changed it again, I would expect there to be 3 entries. Why do we want to act differently here?

Reply:
I was extending the design from the regular edit history that we should only show the most recent event for each file.  I'm not 100% sure it's the right thing for deletes until we have a proper design for it, but it seemed intuitive to me that we'd mainly use this list as a way to restore completely deleted files -- if a file currently exists, the recovery design would probably list the available versions of that file along with the file itself, rather than on some "deleted files" screen, and listing pre-deleted versions of that file would require more extensive changes here, I think.

Either way, I think our different instincts here shows that we should probably do the simplest thing until design weighs in.